src/scancode/cli.py

```diff
@@ -217,724 +225,891 @@ def print_version(ctx, param, value):
+                item_show_func = partial(path_progress_message, verbose=verbose)
+                progress_manager = partial(progressmanager,
+                    item_show_func=item_show_func, verbose=verbose, file=sys.stderr)
+
+            scanners = [Scanner(key='infos', function=get_file_info)]
+            # TODO: add CLI option to bypass cache entirely
+            info_success = scan_codebase(codebase, scanners, processes, timeout,
+                with_timing=timing, with_info=True, progress_manager=progress_manager)
+
+            codebase.timings['collect-info'] = time() - info_start
+            info_is_collected = True
+
+            success = success and info_success
+
+        ############################################################################
+        # 5. run prescans
+        ############################################################################
+
+        pre_scan_start = time()
+        if not quiet and not verbose and pre_scan_plugins:
+            echo_stderr('Run pre-scan plugins...', fg='green')
+
+        # TODO: add progress indicator
+        for name, plugin in pre_scan_plugins.items():
+            plugin_prescan_start = time()
+            if verbose:
+                echo_stderr('Run pre-scan plugin: %(name)s...' % locals(),
+                            fg='green')
+
+            try:
+                plugin.process_codebase(codebase)
+            except:
+                msg = 'ERROR: failed to run pre-scan plugin: %(name)s:' % locals()
+                echo_stderr(msg, fg='red')
+                echo_stderr(traceback.format_exc())
+                ctx.exit(2)
+
+            codebase.update_counts()
+            timing_key = 'prescan_%(name)s' % locals()
+            setup_timings[timing_key] = time() - plugin_prescan_start
+
+        codebase.timings['pre-scan'] = time() - pre_scan_start
+
+        ############################################################################
+        # 6. run scans.
+        ############################################################################
+
+        scan_start = time()
+        scanners = []
+        # add info is requested or needed but not yet collected
+        stages_needs_info = any(p.needs_info for p in
+            (post_scan_plugins.values() + output_plugins.values()))
+
+        with_info = info or stages_needs_info
+        codebase.with_info = with_info
+        if not info_is_collected and with_info:
+            scanners = [Scanner(key='infos', function=get_file_info)]
+
+        scan_sorter = lambda s: (s.sort_order, s.name)
+        for scanner in sorted(scanner_plugins.values(), key=scan_sorter):
+            scanner_kwargs = scanner.get_own_command_options_kwargs()
+            func = scanner.get_scanner(**scanner_kwargs)
+            scanners.append(Scanner(key=scanner.name, function=func))
+
+        if TRACE_DEEP: logger_debug('scancode: scanners:', scanners)
+
+        if scanners:
```

Comment:
Is it possible to run scancode without any actual scanners? Otherwise I'd suggest to change this to `if not scanners: errror...`. If it *is* possible, `scan_names` needs to be initialized outside of the `if`, since it is referenced at the end of the function when creating the summary.

Reply:
In the `develop` branch no scanners means that the `-clip` options will be used. Here I only enabled `--info` if no scans were selected. The scan names are therefore never empty at this stage. They are also created only inside the `if scanners:` at https://github.com/nexB/scancode-toolkit/blob/787-new-plugin-architecture/src/scancode/cli.py#L776
Thank you for your eagle eyes!

Reply:
Right, `scan_names` is only created inside the `if scanners:` block, but it is referenced after the block ends at https://github.com/nexB/scancode-toolkit/blob/787-new-plugin-architecture/src/scancode/cli.py#L857
Are you saying that `scanners` will never be empty? If so, why the `if`? Maybe this is more of a readability issue and not a bug. In that case, don't let me distract you from more important things. :)